### PR TITLE
Recoloring

### LINKED ISSUE #167: Revolutionary Tricolors
The converter happens to convert revolutionary tricolors to a full-white flag.

### PR DESCRIPTION
**Pull request by Zemurin**
Fixes #167
- rewrote color parser, which didn't function at all (colors were being loaded from vanilla country files)
- rewrote custom flag generation, for both custom colors and revolutionary tricolors.
- removed some excess debug spam
- since revolutionary status now forces bourgeois_dictatorship, correct (republic) flag will load, while other (fascist etc) remain unaffected.

flags

### CODE REVIEW

**Comment by Zemurin** [comment]
Green, this is ready for review.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE. */
+
+
+#include "EU4NationalSymbol.h"
+#include "Log.h"
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp]: I don't see any logging, so you can probably remove this.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h [hunk 2]
```diff
@@ -19,28 +19,44 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
-
-#ifndef CUSTOM_FLAG_MAPPER
-#define CUSTOM_FLAG_MAPPER
-
+#ifndef EU4_CUSTOMCOLORS_H_
+#define EU4_CUSTOMCOLORS_H_
 
 
+#include "Color.h"
+#include "Date.h"
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: I don't see date used anywhere, so can probably be removed.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h [hunk 3]
```diff
@@ -19,28 +19,44 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
 #include <memory>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: Can probably be removed.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h [hunk 4]
```diff
@@ -19,28 +19,44 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
-#include <string>
-#include <tuple>
-using namespace std;
+#include <optional>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: Ditto

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h [hunk 5]
```diff
@@ -19,28 +19,44 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+#include <set>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: Ditto

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+THE SOFTWARE. */
+
+
+#ifndef EU4_NATIONALSYMBOL_H_
+#define EU4_NATIONALSYMBOL_H_
+
+
+
+#include "Color.h"
+#include "Date.h"
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: Can probably remove.

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+#include "EU4CustomColors.h"
+#include <memory>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: Ditto.

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+#include <optional>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: Ditto.

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+#include <set>
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: Ditto.

File: EU4toV2/Source/EU4World/EU4Country.cpp [hunk 10]
```diff
@@ -21,6 +21,7 @@ THE SOFTWARE. */
 
 #include "EU4Country.h"
 #include "Country/EU4GovernmentSection.h"
+#include "Country/EU4NationalSymbol.h"
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/EU4Country.cpp]: This is in the .h file, so shouldn't be needed here.

File: EU4toV2/Source/EU4World/EU4Country.cpp [hunk 11]
```diff
@@ -107,27 +105,21 @@ EU4::Country::Country(
...
+			colorColor.RandomlyFlunctuate(30);
 			// Countries whose colors are included in the object here tend to be generated countries,
 			// i.e. colonial nations which take on the color of their parent. To help distinguish 
 			// these countries from their parent's other colonies we randomly adjust the color.
-			color.RandomlyFlunctuate(30);
+			nationalColors.setMapColor(colorColor);
 		}
 	);
-	registerKeyword(std::regex("colors"), [this](const std::string& colorsString, std::istream& theStream)
+	registerKeyword(std::regex("colors"), [this, theVersion](const std::string& colorsString, std::istream& theStream)
```
Comment (Idhrendur) [inline EU4toV2/Source/EU4World/EU4Country.cpp]: Looks like theVersion isn't being used, you can drop it from the []

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+#include "NewParserToOldParserConverters.h"
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp]: done

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
+/*Copyright(c) 2019 The Paradox Game Converters Project
```
[the remark this answers is not in the document]
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp]: done.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp (hunk 1 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.cpp]: done

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h (hunk 2 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: done

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h (hunk 3 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: cant drop color, I need it.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h (hunk 4 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: dropped.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h (hunk 5 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: same

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h (hunk 6 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: done

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h (hunk 7 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: done

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h (hunk 8 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: done

File: EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h (hunk 9 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/Country/EU4NationalSymbol.h]: done

File: EU4toV2/Source/EU4World/EU4Country.cpp (hunk 10 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/EU4Country.cpp]: indeed.

File: EU4toV2/Source/EU4World/EU4Country.cpp (hunk 11 above)
Comment (Zemurin) [reply EU4toV2/Source/EU4World/EU4Country.cpp]: did a version switch first but then changed my mind and forced all to new structure. dropped.

File: EU4toV2/Source/EU4World/Country/EU4CustomColors.h (hunk 3 above)
Comment (Idhrendur) [reply EU4toV2/Source/EU4World/Country/EU4CustomColors.h]: I must have misclicked. I meant date.h

**Pull request merged**